Args is not comptime or const

Code: 

```zig
const positFrom = struct {
    fn from(x: anytype) Posit16 {
      const T = @TypeOf(x);
      return switch( @typeInfo(T) ) {
        .ComptimeInt => Posit16.fromInt(std.math.IntFittingRange(x,x), x),
        .Int => Posit16.fromInt(T, x),
        .Float => Posit16.fromFloat(T, x),
        else => @compileError("Unsupported type"),
      };
    }
  };
  const simpleCtx = math_eval.contexts.simpleCtx({});
  const c = try math_eval.eval("posit(10)", math_eval.contexts.fnMethodCtx(simpleCtx, .{}), .{ .posit = positFrom.from});
  std.debug.print("c = {}\n", .{c});
```

If I use the fnMethodCtx I get the compile error:
```bash
src/math-eval/math-eval/src/eval.zig:153:27: error: variable of type 'eval.EvalExprTupleImpl(.{ .{ .integer = 10 } },contexts.FnMethodCtx(contexts.SimpleCtx(void),.{}),struct{comptime posit: fn(anytype) posit2.Posit(16,1) = (function 'from')})' must be const or comptime
                var args: Args = undefined;
                          ^~~~
src/math-eval/math-eval/src/eval.zig:108:12: note: struct requires comptime because of this field
    return @Type(.{ .Struct = .{
           ^~~~~
src/math-eval/math-eval/src/eval.zig:29:20: note: called from here
    return evalImpl(root, ctx, inputs);
           ~~~~~~~~^~~~~~~~~~~~~~~~~~~
src/main.zig:66:35: note: called from here
      const c = try math_eval.eval("posit(10)", math_eval.contexts.fnMethodCtx(simpleCtx, .{}), .{ .posit = positFrom.from});
                    ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

However if I use simpleCtx It works
I tried to add comptimes all over the place, but I couldn't fix it. Maybe upstream bug?

--- a/build.zig
+++ b/build.zig
@@ -5,20 +5,13 @@ pub fn build(b: *Build) void {
     const util_mod = b.createModule(.{
         .source_file = Build.FileSource.relative("src/util.zig"),
     });
-    const eval = b.addModule("eval", .{
+    const eval_mod = b.addModule("eval", .{
         .source_file = Build.FileSource.relative("src/eval.zig"),
         .dependencies = &.{
             .{ .name = "util", .module = util_mod },
         },
     });
-    const contexts = b.addModule("contexts", .{
-        .source_file = Build.FileSource.relative("src/contexts.zig"),
-        .dependencies = &.{
-            .{ .name = "util", .module = util_mod },
-            .{ .name = "eval", .module = eval },
-        },
-    });
-    _ = contexts;
+    _ = eval_mod;
 
     const target = b.standardTargetOptions(.{});
     const optimize = b.standardOptimizeOption(.{});
@@ -31,16 +24,6 @@ pub fn build(b: *Build) void {
     const run_main_tests = b.addRunArtifact(main_tests);
     main_tests.addModule("util", util_mod);
 
-    const contexts_tests = b.addTest(.{
-        .root_source_file = Build.FileSource.relative("src/contexts.zig"),
-        .target = target,
-        .optimize = optimize,
-    });
-    const run_contexts_tests = b.addRunArtifact(contexts_tests);
-    contexts_tests.addModule("eval", eval);
-    contexts_tests.addModule("util", util_mod);
-
     const test_step = b.step("test", "Run library tests");
     test_step.dependOn(&run_main_tests.step);
-    test_step.dependOn(&run_contexts_tests.step);
 }

--- a/src/contexts.zig
+++ b/src/contexts.zig
@@ -132,6 +132,7 @@ pub fn SimpleCtx(comptime SubCtx: type) type {
                     return ctx.sub_ctx.evalFuncCall(callee, args);
                 }
             }
+            return @call(.auto, callee, args);
         }
 
         pub fn EvalUnOp(comptime op: UnOp, comptime T: type) type {
@@ -555,6 +556,10 @@ test fnMethodCtx {
     const CustomNum = enum(i32) {
         _,
 
+        pub inline fn from(value: anytype) @This() {
+            return @enumFromInt(value);
+        }
+
         pub inline fn add(self: @This(), other: @This()) @This() {
             return @enumFromInt(@intFromEnum(self) + @intFromEnum(other));
         }
@@ -579,13 +584,16 @@ test fnMethodCtx {
         .@"*" = "mul",
         .evalFuncCall = "mul",
     });
-    try util.testing.expectEqual(@as(CustomNum, @enumFromInt(2)), eval.eval("a + -b - c", fm_ctx, .{
+    try util.testing.expectEqual(CustomNum.from(2), eval.eval("a + -b - c", fm_ctx, .{
         .a = @as(CustomNum, @enumFromInt(22)),
         .b = @as(CustomNum, @enumFromInt(9)),
         .c = @as(CustomNum, @enumFromInt(11)),
     }));
-    try util.testing.expectEqual(@as(CustomNum, @enumFromInt(77)), eval.eval("a(b) * 1", fm_ctx, .{
+    try util.testing.expectEqual(CustomNum.from(77), eval.eval("a(b) * 1", fm_ctx, .{
         .a = @as(CustomNum, @enumFromInt(11)),
         .b = @as(CustomNum, @enumFromInt(7)),
+    }));
+    try util.testing.expectEqual(CustomNum.from(62), eval.eval("num(31)(2)", fm_ctx, .{
+        .num = CustomNum.from,
     }));
 }

--- a/src/eval.zig
+++ b/src/eval.zig
@@ -93,11 +93,16 @@ fn EvalExprTupleImpl(
     var fields: [list.len]std.builtin.Type.StructField = undefined;
     for (&fields, list, 0..) |*field, arg, i| {
         const T = EvalImpl(arg, Ctx, Inputs);
+        const comptime_only = util.typeIsComptimeOnly(T) orelse @compileError("Cannot have an argument of type " ++ @typeName(T));
+
         field.* = .{
             .name = std.fmt.comptimePrint("{d}", .{i}),
             .type = T,
-            .default_value = null,
-            .is_comptime = false,
+            .default_value = if (!comptime_only) null else blk: {
+                const val: T = evalImpl(arg, @as(Ctx, undefined), @as(Inputs, undefined)) catch |err| @compileError(@errorName(err));
+                break :blk &val;
+            },
+            .is_comptime = comptime_only,
             .alignment = 0,
         };
         if (@sizeOf(T) == 0) {
@@ -150,11 +155,21 @@ inline fn evalImpl(
 
             const Args = EvalExprTupleImpl(fc.args, Ctx, Inputs);
             const args: Args = args: {
-                var args: Args = undefined;
-                inline for (fc.args, 0..) |arg, i| {
-                    args[i] = try evalImpl(arg, ctx, inputs);
+                if (util.typeIsComptimeOnly(Args).?) {
+                    var args: Args = undefined;
+                    inline for (fc.args, 0..) |arg, i| {
+                        if (@typeInfo(Args).Struct.fields[i].is_comptime) continue;
+                        args[i] = try evalImpl(arg, ctx, inputs);
+                    }
+                    break :args args;
+                } else comptime {
+                    var args: Args = undefined;
+                    inline for (fc.args, 0..) |arg, i| {
+                        if (@typeInfo(Args).Struct.fields[i].is_comptime) continue;
+                        args[i] = try evalImpl(arg, ctx, inputs);
+                    }
+                    break :args args;
                 }
-                break :args args;
             };
 
             break :blk ctx.evalFuncCall(callee, args);

--- a/src/util.zig
+++ b/src/util.zig
@@ -197,3 +197,42 @@ pub fn GetPayloadIfErrorUnion(comptime T: type) type {
         else => T,
     };
 }
+
+pub inline fn typeIsComptimeOnly(comptime T: type) ?bool {
+    comptime return switch (@typeInfo(T)) {
+        .Type => true,
+        .Void => false,
+        .Bool => false,
+        .NoReturn => null,
+        .Int => false,
+        .Float => false,
+        .Pointer => |pointer| if (typeIsComptimeOnly(pointer.child)) |is_comptime_only| is_comptime_only and switch (pointer.size) {
+            .One => @typeInfo(pointer.child) != .Opaque or (pointer.is_const and @typeInfo(pointer.child) != .Fn),
+            else => true,
+        } else null,
+        .Array => |array| array.len != 0 and typeIsComptimeOnly(array.child),
+        .Struct => |structure| for (structure.fields) |field| {
+            if (field.is_comptime) continue;
+            if (!(typeIsComptimeOnly(field.type) orelse return null)) continue;
+            break true;
+        } else false,
+        .ComptimeFloat => true,
+        .ComptimeInt => true,
+        .Undefined => null,
+        .Null => null,
+        .Optional => |optional| typeIsComptimeOnly(optional.child),
+        .ErrorUnion => |optional| typeIsComptimeOnly(optional.payload),
+        .ErrorSet => false,
+        .Enum => |enumeration| typeIsComptimeOnly(enumeration.tag_type),
+        .Union => |@"union"| typeIsComptimeOnly(@"union".tag_type orelse enum {}) and for (@"union".fields) |field| {
+            if (!typeIsComptimeOnly(field.type)) continue;
+            break true;
+        } else false,
+        .Fn => true,
+        .Opaque => null,
+        .Frame => false,
+        .AnyFrame => false,
+        .Vector => |vector| typeIsComptimeOnly(vector.child),
+        .EnumLiteral => true,
+    };
+}

